config.cc

```diff
@@ -2221,7 +2256,19 @@ QString getPidFileName() THROW_SPEC( exError )
 
 QString getHistoryFileName() THROW_SPEC( exError )
 {
+  #ifdef Q_OS_LINUX
+    QDir home = getHomeDir();
+    if (home.dirName().endsWith(".goldendict"))
+    {
+        return getHomeDir().filePath( "history" );
+    }
+    else
+    {
+        return getDataDir().filePath( "history" );
+    }
+  #else
   return getHomeDir().filePath( "history" );
+  #endif
```

Comment:
This code can be shortened by reusing the existing `return getHomeDir().filePath( "history" );` line as a fallback if the `#if` and `if()` branches above are not taken.